### DISCUSSION TITLE
Error: Circular structure in "getInitialProps" result of page _app.js

### DISCUSSION THREAD

**Discussion by namlq93** [vercel/next.js#17403] [Help] [answered]
I want to pass `ctx` inside MyApp.getInitialProps to MyApp, and in MyApp I will use ctx to handle redirect if user doesn't have admin role.

MyApp.getInitialProps:

```
    const initialProps = await App.getInitialProps(ctx);
    const pageProps = { ...initialProps.pageProps, userLogin, ctx };

    return { pageProps: { ...pageProps } };
```

MyApp: 

```
const MyApp = ({ Component, pageProps }) => {
    console.log(pageProps);
    const { userLogin, ctx } = pageProps;
    if (pageProps.isAdmin && !userLogin?.hasCustomerAdminRole) {
        console.log('admin');
        if (ctx.res) {
            ctx.res.writeHead(303, {
                Location: `/login?redirectPath=${ctx.req.url}`,
            });
        }
    }
    return (
        <>
            ....
        </>
    );
};
```

![Screen Shot 2020-09-28 at 13 29 28](https://user-images.githubusercontent.com/41458989/94397840-b2b32200-018e-11eb-9e2f-9831b71792ad.png)

On browser, I've got `500 (Internal Server Error)`

How to fix it, thank you so much.

**Comment by namlq93** [comment, marked as answer]
I fixed it, check inside MyApp.getInitialProps

**Reply by samynaj** [reply]
hi pls can you share how you solved this issue
